[Tooling] Re-implement AppLocalizedStrings linter in swift

AliSoftware opened the pull request:
This work is part of paaHJt-2Ib-p2, and a follow up of the annoucement in paNNhX-nP-p2.

## What

This re-implements the `AppLocalizedString` linter introduced in https://github.com/wordpress-mobile/WordPress-iOS/pull/18427 (which was written in Ruby) to re-write it in pure Swift

## Why

 - Because it's always nice to have one less tooling depending on Ruby
 - Because one of our mid-to-long term plan is to not require developers to install Ruby and `rbenv` on their Mac anyway (and even if we're not there yet, better not continue increasing our ruby usage for devs if possible)
 - Because I couldn't resist being nerd-snipe'd by @mokagio in https://github.com/wordpress-mobile/WordPress-iOS/pull/18427#pullrequestreview-954314088

## How

 - Since `project.pbxproj` files are in the ASCII-Plist format, they can easily be parsed and decoded using `PropertyListDecoder`
 - I thus implemented a couple of `Decodable` types to interpret the various structures present in a `.pbxproj` file (only the ones relevant to our needs)
 - The _main_ part of the swift script then use those `Decodable` models to parse the `.xcodeproj` file, iterate on its `PBXNativeTarget` objects of type `.appExtension`, find their build phases, resolve the path of each `PBXBuildFile` listed as child of the "Compile Sources" phase, and lint each of them in turn.
 - The Build Phase calling this linter script (~in each of the app extension target of the Xcode project~) calls a `LintAppLocalizedStringsUsage.sh` wrapper script which takes care of:
   - Pre-compiling the swift script into a binary (using the proper macOS SDK), to speed up further executions (instead of having to re-interpret the swift script each time)
   - Run the pre-compiled script on the appextension targets

## To Test

 - Build the project, validate it doesn't get any error while trying to run the lint Script Build Phase nor does trigger any violation
 - Optional: Go to the Build log, search for `[Lint]` to find all the instances invoking the linter script (there should be one invocation per app extension target), and verify that while the first invocation might have taken a couple of seconds, all the successive invocations used the pre-compiled version of the linter and took <1s
 - Introduce a violation on purpose, by replacing any call of `AppLocalizedString` in the codebase (in a file used by an app extension) by `NSLocalizedString`
 - Build the project again, and check that the violation was flagged as an error in your IDE
 - fix the violation you introduced on purpose, then build the project again, and check the error disappeared

#### Optional:
 - Go to the Build log, search for `[Lint]` and verify that all the invocations of the linter were fast and used the pre-compiled version of the linter.
 - Edit the `Scripts/BuildPhases/LintAppLocalizedStringsUsage.swift` file to add a space somewhere (so it is seen as changed)
 - Build the project again, and verify that the first invocation of the script in a `[Lint]` phase detected the source of the linter changed and re-compiled it from the new source before proceeding

AliSoftware committed:
Introduce Swift version of the AppLocalizedString linter

AliSoftware committed:
Call the new swift linter script in pbxproj

Using a .sh wrapper script to force using the macosx SDK when running the linter swift script, otherwise it will assume iphoneos/iphonesimulator based on the target running the script

AliSoftware committed:
Cache pre-compiled version of the script

To speed up the invocations

AliSoftware committed:
Clean up old ruby scripts

AliSoftware commented on Scripts/BuildPhases/LintAppLocalizedStringsUsage.sh:
```diff
@@ -0,0 +1,17 @@
+  echo "Pre-compiled linter script ready"
+fi
+
+"$LINTER_EXEC" "${PROJECT_FILE_PATH:-$1}" "${TARGET_NAME:-$2}"
```
ℹ️  A previous version of this (ded9b82f07a877b712fa27814fd8a8ded52a90b6) invoked the `.swift` script directly (i.e. using the swift interpreter, without having to pre-compile the script into a program), which worked… but each invocation took ~3s (the time to spawn the Swift interpreter, parse/typecheck/interpret the swift script, and finally run it, I guess), at least according to the timings reported in Xcode's build logs UI.

Which is why I decided to try this trick to pre-compile the `.swift` script into a binary on first invocation, and run the pre-compiled binary on subsequent invocations. It has indeed led to nice speedups when running the Build Phases in Xcode (roughly: 4s on first invocation by the first target needing it… but then 0.5s on any subsequent runs 🎉 )

I also considered just commit the pre-compiled version of the script as a binary into the repo, but decided against that in fear of compatibility issues (Intel vs M1 binaries, issues when resolving dependencies of the compiled script with `.dylib` Swift runtime libraries…) and thought it would be simpler to compile it JIT on the user's machine instead like here.

---

PS: The pre-compiled version of the script is saved in `${BUILD_DIR}` (e.g. `<Path to DerivedData>/WordPress-<someuuid>/Build/Products/LintAppLocalizedStringsUsage`), so it will be cleaned alongside any other compiled product of your project if you do a Clean Build (or just Clean Build Products folder)

AliSoftware committed:
Rearrange the swift code and comment it

To make it more readable/understandable/maintainable with a nicer structure and more detailed explanations

AliSoftware commented on Scripts/BuildPhases/LintAppLocalizedStringsUsage.swift:
```diff
@@ -0,0 +1,324 @@
+        }
+    }
+
+    /// Helper type to ensure the `isa` field of a `PBXObject` contains the name of the expected type to decode as its value.
+    struct ISA<T>: RawRepresentable, Decodable, CustomDebugStringConvertible {
+        var rawValue: String
+        init?(rawValue: String) {
+            guard rawValue == String(describing: T.self) else { return nil }
+            self.rawValue = rawValue
+        }
+        var debugDescription: String { self.rawValue }
+    }
```
ℹ️  This is used in the various `struct XYZ: PBXObject` types as a trick to ensure their `let isa: ISA<Self>` field only allow the value corresponding to `T`'s type name when decoding those objects.

AliSoftware commented on Scripts/BuildPhases/LintAppLocalizedStringsUsage.swift:
```diff
@@ -0,0 +1,324 @@
+    /// Represents a group (aka "folder") contained in the project tree.
+    struct PBXGroup: PBXReference {
+        enum ISA: String, Decodable { case PBXGroup, XCVersionGroup, PBXVariantGroup }
+        private let isa: ISA
```
ℹ️ We don't use the `ISA<Self>` trick on this one because we want multiple `isa` values (`"PBXGroup"`, `"XCVersionGroup"`, `"PBXVariantGroup"`) to all be allowed and all decode as a `PBXGroup` instance.

Alternatively, I could have created 3 separate `struct PBXGroup`, `struct XCVersionGroup` and `struct PBXVariantGroup`, each using `ISA<Self>` like the other `PBXObject` types. But that would have required to duplicate the same list of fields 3 times, while in our specific scenario and use case we don't care about any field that would otherwise make those 3 types differ, and only care about their 4 properties listed below for our usage, so it was not worth duplicating each type.

Another route I tried was to use inheritance instead, but that opened its own can of worms (like having to implement a `required init` ourselves unlike when using `struct` where we can rely on the auto-generated one, not being able to use `ISA<Self>` if `Self` (aka the class in question) is not `final`, …), so I decided to keep `struct` and value types all the way and use this trick instead after all.

mokagio replied:
> ℹ️ We don't use the `ISA<Self>` trick on this one because we want multiple `isa` values (`"PBXGroup"`, `"XCVersionGroup"`, `"PBXVariantGroup"`) to all be allowed and all decode as a `PBXGroup` instance.

Do you think it would be beneficial to leave this in the code so that future readers know about it?

```suggestion
        enum ISA: String, Decodable { case PBXGroup, XCVersionGroup, PBXVariantGroup }
        // We don't have a `ISA<Self>` here because we want multiple `isa` values to all be allowed and all decode as a `PBXGroup` instance (`"PBXGroup"`, `"XCVersionGroup"`, `"PBXVariantGroup"`)
        private let isa: ISA
```

---

As for the different approaches... I think this is fine. This is a self contained script so we don't need to have it _perfect_ or worry about extensibility and tradeoffs too much 😄 👍

AliSoftware replied:
Actually the alternate approach I have in my backback (but still haven't found the time in my crazy MIE schedule to apply and push… so by now we should probably merge the PR as is and I'll do a follow-up PR to apply my new approach afterwards on top of it, I guess) don't use the `ISA<Self>` trick anymore.

Instead with the new approach I found a way to decode the `isa` field and then re-decode the same `KeyedContainer` but this time with the right class (`switch`-ing on the `isa` value to decode it), all in one go, i.e. without having to loop over all the possible `Decodable` classes until one succeeds.

In that approach I have in mind, the `isa` is a `String` and then each `PBXObject` type conforms to a protocol which has a `static func match(isa: String) -> Bool` requirement with a default implementation of `static func match(isa: String) -> Bool { isa == String(describing: self) }`. Then all my structs representing `PBXObjects` use that default implementation, except `PBXGroup` which provides an explicit implementation doing `["PBXGroup", "PBXVariantGroup", "XCVersionGroup"].contains(isa)`.

I could still provide a comment about this even in the new approach (if I get to push it one day…), but I think the new approach will make it more understandable (not having the reader of the code understand how the `ISA<Self>` trick works) and might thus make the comment less necessary…

AliSoftware replied:
Comment added via your suggestion in [9835f03](https://github.com/wordpress-mobile/WordPress-iOS/pull/18462/commits/9835f037667cd6209349944e06188924cd0a5a11)

AliSoftware commented on Scripts/BuildPhases/LintAppLocalizedStringsUsage.swift:
```diff
@@ -0,0 +1,324 @@
+
+    /// Wrapper helper to decode any `PBXObject` based on the value of their `isa` field
+    @propertyWrapper
+    struct PBXObjectWrapper: Decodable, CustomDebugStringConvertible {
+        let wrappedValue: PBXObject
+        static let knownTypes: [PBXObject.Type] = [
+            PBXProject.self,
+            PBXGroup.self,
+            PBXFileReference.self,
+            PBXNativeTarget.self,
+            PBXSourcesBuildPhase.self,
+            PBXBuildFile.self
+        ]
+
+        init(from decoder: Decoder) throws {
+            // Try to decode each known types in turn, until we find one that succeeds decoding — by having the expected `isa` field value.
+            for objectType in Self.knownTypes {
+                if let object = try? objectType.init(from: decoder) as PBXObject {
+                    self.wrappedValue = object
+                    return
+                }
+            }
+            self.wrappedValue = try UnknownPBXObject(from: decoder) as PBXObject // Fallback
+        }
+        var debugDescription: String { String(describing: wrappedValue) }
+    }
```
ℹ️  Inspired by https://forums.swift.org/t/decoding-to-different-models-based-on-json-key/19288

I tried other approaches to solve the "decode the right subtype based on a field/key (i.e. `isa`), but this ended up being the simpler and less convoluted one, and also the most performant (especially compared to the other I tried of decoding the dictionary twice — first one to just extract the `isa` and determine the proper type, and another one to reparse the entire dictionary to instantiate said type — which slowed down considerably the parsing, at least when I tested it in my Swift Playground)

AliSoftware replied:
Actually I couldn't help myself but play with Swift over the weekend to try other ideas, and found one that I think might look neater. I might update the code to use it at some point, either in this PR or in a next one.

mokagio replied:
Looking forward to it 👍

AliSoftware commented on Scripts/BuildPhases/LintAppLocalizedStringsUsage.swift:
```diff
@@ -0,0 +1,324 @@
+
+        /// The list of `PBXBuildFile` build file references included in a given target.
+        func buildFiles(for target: PBXNativeTarget) -> [PBXBuildFile] {
+            guard let sourceBuildPhase: PBXSourcesBuildPhase = target.buildPhases.lazy.compactMap(object(id:)).first else { return [] }
```
ℹ️  The magic of Swift type system: ✨ 

 - `sourceBuildPhase` being typed as `PBXSourceBuildPhase`…
 - … makes the type system infer that the whole `….compactMap(…).first` expression must be of type `PBXSourceBuildPhase?`…
 - … which means the result of `….compactMap(…)` must be of type `[PBXSourceBuildPhase]`…
 - … which implies that the function called inside the `compactMap` must return a `PBXSourceBuildPhase?` (and `compactMap` will only keep the returned values which are not `nil`)…
 - … which finally means that the generic type `T` for the `object(id:)` method specialization can be infered to be `T = PBXSourcesBuildPhase`…

This is also why, btw, this `compactMap` code filters to only go thru the `PBXSourcesBuildPhase` instances, ignoring / filtering out any potential other type of Build Phase that could be present in the `target.buildPhase` array 🙂 ✨

mokagio replied:
Beautiful 🤩

AliSoftware commented on Scripts/BuildPhases/LintAppLocalizedStringsUsage.swift:
```diff
@@ -0,0 +1,324 @@
+    /// Semantic type for strings that correspond to an object' UUID in the `pbxproj` file
+    typealias ObjectUUID = String
+
+    /// Builds an `Xcodeproj` instance by parsing the an `.xcodeproj` or `pbxproj` file at the provided URL
```
```suggestion
    /// Builds an `Xcodeproj` instance by parsing the `.xcodeproj` or `.pbxproj` file at the provided URL.
```

AliSoftware committed:
Fix typos in code comments

AliSoftware committed:
Run linter on app targets instead of appex ones

This changes the strategy to only have a single Script Build phase on the parent app target, rather than having one on each app extension target.

The initial rationale for calling the linter on each app extension target independently (and limiting the linter just to that target) was that it would ensure the linter would be run even if we only try to build the app extension target independently (e.g. selecting the scheme associated with that app extension when building). SO I thought it made more sense to make each target call the linter for their own.

But in practice, an app extension is never build / run without building the app target as well. Especially, if you look at each scheme corresponding to an app extension (e.g. the `WordPressStatsWidgets` scheme) you can see in the "Build" tab of the scheme that it will also build the `WordPress` app target anyway.

Given this, it ends up being (1) easier to maintain (e.g. if we add new appextension targets in the future), and (2) more performant / faster to just run the linter swift script once, from the app target, and have it scan the files for each of the appextension targets it finds in the `xcodeproj` in one go — rather than running it from scratch as many times as there are appextension targets in the project.

The performance boost we'll get from this new approach is because each invocation of the linter parses the whole `project.pbxproj` file (to find the list of targets then the list of files built by the targets). So only doing that once when we build the `WordPress` or `Jetpack` targets is more efficient than re-parsing the project file multiple times for each app extension target being built.

AliSoftware commented:
Update: I've changed the strategy used to call the linter in the Build Phases, to **invoke the linter only once** from the app targets (`WordPress` and `Jetpack`), and let them scan for build files in each app extension target in one go — instead of the previous strategy of invoking the linter from each app extension target (and only scanning for build files for that target). cc @mokagio 

The end result in error reporting is the same (the linter still loops on each app extension target, then lint each file being built in said app extension target, leading to the exact same errors being reported, including a violation being reported as many times as there are app extension targets it would cause issues for).

But the benefit of the approach is that it will only make the linter parse the `project.pbxproj` file once (then looping on all the app extension targets after that single parsing operation), as opposed to parsing as many times as there are app extension targets. See also 826bd0a6a5db7352b3298fcca9370b73510b1902 for the rationale

mokagio commented:
@AliSoftware moving this to 19.9.

I started reviewing and testing to merge before the code freeze and close the loop, but that's taking a while and I don't want to further delay the code freeze.

Since this is a tooling PR, whether or not it lands in 19.8 or 19.9 doesn't affect our end users.

AliSoftware commented on Scripts/BuildPhases/LintAppLocalizedStringsUsage.swift:
```diff
@@ -0,0 +1,324 @@
+    ///   up into the chain of parent `PBXGroup` containers to construct the successive relative paths of groups using `sourceTree = "<group>"`
+    private lazy var referrers: [ObjectUUID: ObjectUUID] = {
+        var referrers: [ObjectUUID: ObjectUUID] = [:]
+        func recurseIfGroup(objectID: ObjectUUID, indent: String = "") {
+            guard let group = try? (self.pbxproj.object(id: objectID) as PBXGroup) else { return }
+            for childID in group.children {
+                referrers[childID] = objectID
+                recurseIfGroup(objectID: childID, indent: "\(indent)  ")
+            }
```
TODO: `indent` parameter is debug leftover, should remove it

AliSoftware commented on Scripts/BuildPhases/LintAppLocalizedStringsUsage.swift:
```diff
@@ -0,0 +1,324 @@
+
+        switch object.sourceTree {
+        case .absolute:
+            return URL(fileURLWithPath: object.path!)
```
A `PBXReference` with `sourceTree = absolute` but no `path` would make no sense, hence the force unwrap.

Should I catch potential pbxproj inconsistencies like that and make it throw instead of `!`? Or maybe at least a `fatalError(message)`…

AliSoftware commented on Scripts/BuildPhases/LintAppLocalizedStringsUsage.swift:
```diff
@@ -0,0 +1,324 @@
+        case .projectRoot:
+            return object.path.map { URL(fileURLWithPath: $0, relativeTo: projectDirectory) } ?? projectDirectory
+        case .buildProductsDir, .devDir, .sdkDir:
+            fatalError("Unsupported relative reference (relative to: \(object.sourceTree)")
```
We don't (and shouldn't) have files in our pbxproj referenced relative to SDK or BuildProducts or Dev Dir, so that's not a case we will encounter nor need to implement right now…

But maybe I should make this throw instead, and just print a warning + skip any file found in the pbxproj that might end up using this in the future?

AliSoftware reviewed (commented):
Some self-review after a good night sleep + re-read of my own code. TL;DR:
 - I could replace some force unwraps with throws and/or `fatalError(clearMessage)`
 - Thought about a (better?) alternative solution for the "decode object from dict based on isa field from same dict" decoding trick

I'm ok with either:

1. Landing this PR as is (to avoid `.pbxproj` changes to drift too much and get conflicts) since it does the job, and do a follow-up PR to fix force-unwraps and inconsistent pbxproj error handling + and try alternative idea for `isa`-decoding
2. Or keep this PR open and don't land it until I make those changes and re-request another review, in case you prefer we only land it once swift code is looking super-clean

crazytonyli commented on Scripts/BuildPhases/LintAppLocalizedStringsUsage.sh:
```diff
@@ -0,0 +1,17 @@
+
+if [ ! -x "${LINTER_EXEC}" ] || ! (shasum -c "${LINTER_EXEC}.shasum" >/dev/null 2>/dev/null); then
+  echo "Pre-compiling linter script to ${LINTER_EXEC}..."
+  swiftc -sdk "$(xcrun --sdk macosx --show-sdk-path)" "${SCRIPT_SRC}" -o "${LINTER_EXEC}"
```
Do you think the shell script should exit with error when failing to compile the Swift script? I wouldn't be surprised if Swift 6 or something breaks source code compatibility.

AliSoftware replied:
That's a good point. The shell script is called by the Script Build Phase using `/bin/bash -eu` And the `-e` flag would make the script exit on first error.

But I'm not sure that flag carries over when the script from the Script Build Phase itself calls this `LintAppLocalizedStringsUsage.sh` one. Maybe I should add `-eu` to the shebang of this file as well, just to be sure.

AliSoftware replied:
Addressed in 2ea977a2ee

crazytonyli commented on Scripts/BuildPhases/LintAppLocalizedStringsUsage.swift:
```diff
@@ -0,0 +1,324 @@
+for target in targetsToLint {
+    let files: [Xcodeproj.PBXBuildFile] = project.buildFiles(for: target)
+    print("Linting the Build Files for \(target.name):")
+    for file in files {
```
I don't image performance is an issue for this script, but I feel like [task group](https://docs.swift.org/swift-book/LanguageGuide/Concurrency.html#ID642) is perfect for this use case here, which potentially can speed up this script a little bit. I'm not sure if it's feasible to use the structured concurrency in a standalone script though...

AliSoftware replied:
Yeah interesting thought, but I'm not sure it's worth it for this context.

That being said… I've been eager to play with structured concurrency and Task Groups (I haven't had the occasion to do much Swift in a while… and in fact working on this Swift implementation of the linter was a great excuse to get back the pleasure of doing some Swift again) so maybe that would be a good excuse to look at this in an future PR to improve on it if I find the time… 🙂

crazytonyli replied:
No worries, I don't think there is really much practical gain in using it here, other than trying some new shiny tech.

AliSoftware replied:
@crazytonyli For the record: I actually tried using `TaskGroup` and it worked but:

 - Because you can't call `async` code from a non-async context:
    - I had to wrap that "main" part of the script inside a `Task { … }` block
    - … and then I had to use a `DispatchSemaphore` (that I'd `signal` at the end of the `Task { … }`) so that I could block the main thread and wait until all tasks were done. It works, but felt a bit dirty to have to rely on a semaphore… 🤷 
 - It messed up the order of the lines I `print`-ed in the `stdout`, obviously because I parsed my files and thus wrote my logs in an async order now.
    - That wouldn't have been too strange if those logs were just a big flat list, but since I print a tree structure, i.e. list of targets then list of files scanned in that target, that messed things up
    - Sure I could have localized the parallelization per target only, and/or actually use the _structured_ part of structured concurrency to create some task dependencies… but I didn't go that far
    - (In a similar vein I thought about using `AsyncSequence` to read the `url.lines` when linting each file, but didn't spend enough time on this to get to that)
 - And ultimately, at least according to my crude estimation of invoking my script using `time <myscript>` a couple of time and averaging the results for both before and after… it appeared that using structured concurrency took… more time overall!! I guess it's the cost of creating threads under the hood for so many small tasks that don't overcome the benefit we'd otherwise have with the parallelisation.

That was still a really nice and fun thing to try, and good excuse to _finally_ play with `TaskGroup`, `async`/`await`, and structured concurrency… so thanks for the suggestion, I had fun anyway 😄

crazytonyli replied:
> it appeared that using structured concurrency took… more time overall!!

🤣 this is awesome. Now I learned the switching thread context may indeed have negative impact on performance. 

Thanks for trying it and sharing the learnings! 👍

mokagio commented on WordPress/WordPress.xcodeproj/project.pbxproj:
```diff
@@ -16636,89 +16629,7 @@
-			shellScript = "\"$SRCROOT/../Scripts/BuildPhases/runRubyScript\" \"LintAppLocalizedStringsUsage.rb\" \"${TARGET_NAME}\"\n";
-			showEnvVarsInLog = 0;
-		};
-		09DBEA4F281333530019724E /* [Lint] Check AppLocalizedString usage */ = {
```
I love how this latest iteration reduced the injection point of the build phase!

AliSoftware replied:
:nod: 🙂

crazytonyli reviewed (commented):
Feel free to discard my comments, I just thought this is an interesting PR to look into 😄

mokagio reviewed (approved):
Great work as always @AliSoftware! I appreciate how you worked on this to not increase our reliance on Ruby, which we'd like to eventually remove for product developers 🙇‍♂️ 

I tested this on my Intel Mac Book pro on both WordPress and Jetpack. It worked as expected, detecting violations and removing the build error after they'd been fixed.

<img width="1506" alt="image" src="https://user-images.githubusercontent.com/1218433/166605498-e7047f99-65af-4db5-91f1-f3b821579ddc.png">

Performance wise. It took 2.5s on the first run and 1s on the subsequent ones. I think this is plenty acceptable for the time being 👍 The difference is due, I'm pretty sure, to the script compilation, which is not something we'd do only on clean builds and anyways takes about a second. 👍 

<img width="664" alt="image" src="https://user-images.githubusercontent.com/1218433/166605454-5cb1f059-91ad-4853-bff9-d45f96a60c78.png">

<img width="511" alt="image" src="https://user-images.githubusercontent.com/1218433/166605465-b77ca819-95e4-4442-8922-0be3a34e33db.png">

AliSoftware committed:
Add comment to explain implementation detail

AliSoftware committed:
Fix shebang of the .sh wrapper script

As per a good point and suggestion from @crazytonyli in https://github.com/wordpress-mobile/WordPress-iOS/pull/18462/files/44d97955be882808c9697ee32804ad19c406210f\#r864327410

AliSoftware reviewed (commented):
Thanks @crazytonyli @mokagio for the review!

I think I'll now merge it, especially to avoid the risk of having the `WordPress.xcodeproj/project.pbxproj` file drift too far from the one in `trunk` and risking conflicts.

But I still hope to create a follow-up PR later this sprint to address the remaining points in the `.swift` script, namely:
 - implement my alternate—and arguably better—implementation for decoding object type based on `isa`
 - replace force-unwraps with nicer errors (for the unlikely cases of encountering inconsistent/corrupted `pbxproj` files, but also to [save some ponies](https://alisoftware.github.io/swift/2015/09/06/thinking-in-swift-1/))

mokagio commented:
> I think I'll now merge it, especially to avoid the risk of having the WordPress.xcodeproj/project.pbxproj file drift too far from the one in trunk and risking conflicts.

Good choice.

AliSoftware commented:
Finally found some time to submit my improvements as follow-up PR here: https://github.com/wordpress-mobile/WordPress-iOS/pull/18525
